stdlib/LinearAlgebra/src/matmul.jl

```diff
@@ -784,18 +784,26 @@ Base.@constprop :aggressive generic_matmatmul!(C::AbstractVecOrMat, tA, tB, A::A
     if BxN != CxN
         throw(DimensionMismatch(lazy"matrix B has axes ($BxK,$BxN), matrix C has axes ($CxM,$CxN)"))
     end
-    if iszero(_add.alpha) || isempty(A) || isempty(B)
-        return _rmul_or_fill!(C, _add.beta)
-    end
-    a1 = first(AxK)
-    b1 = first(BxK)
-    @inbounds for i in AxM, j in BxN
-        z2 = zero(A[i, a1]*B[b1, j] + A[i, a1]*B[b1, j])
-        Ctmp = convert(promote_type(R, typeof(z2)), z2)
-        for k in AxK
-            Ctmp = muladd(A[i, k], B[k, j], Ctmp)
+    if sizeof(R) ≤ 16
```

Comment:
```suggestion
    if sizeof(R) ≤ 16 && !((A isa Adjoint || A isa Transpose) && !(B isa Adjoint || B isa Transpose))
```
If `A` is transposed and `B` is not, the other path is still probably faster for small eltypes, although it'd probably want an `@simd for k in AxK` in that case.

Reply:
I ran the benchmark problems with all 4 combinations regarding transposition of A and/or B for the mixed small eltype case. We can significantly improve all case (as shown in https://github.com/JuliaLang/julia/pull/52038#issuecomment-1801725637) except for the both transposed case. That one seems is clearly faster with the old tiling code:

```julia
julia> @benchmark mymul1!($C, $A, $B, $_add)
BenchmarkTools.Trial: 2 samples with 1 evaluation.
 Range (min … max):  2.972 s …   3.013 s  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     2.992 s              ┊ GC (median):    0.00%
 Time  (mean ± σ):   2.992 s ± 29.576 ms  ┊ GC (mean ± σ):  0.00% ± 0.00%

  █                                                       █  
  █▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁█ ▁
  2.97 s         Histogram: frequency by time        3.01 s <

 Memory estimate: 0 bytes, allocs estimate: 0.

julia> @benchmark mymul2!($C, $A, $B, $_add)
BenchmarkTools.Trial: 2 samples with 1 evaluation.
 Range (min … max):  3.749 s …   3.782 s  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     3.765 s              ┊ GC (median):    0.00%
 Time  (mean ± σ):   3.765 s ± 23.857 ms  ┊ GC (mean ± σ):  0.00% ± 0.00%

  █                                                       █  
  █▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁█ ▁
  3.75 s         Histogram: frequency by time        3.78 s <

 Memory estimate: 0 bytes, allocs estimate: 0.

julia> @benchmark mymul3!($C, $A, $B, $_add)
BenchmarkTools.Trial: 2 samples with 1 evaluation.
 Range (min … max):  3.735 s …  3.739 s  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     3.737 s             ┊ GC (median):    0.00%
 Time  (mean ± σ):   3.737 s ± 2.920 ms  ┊ GC (mean ± σ):  0.00% ± 0.00%

  █                                                      █  
  █▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁█ ▁
  3.73 s        Histogram: frequency by time        3.74 s <

 Memory estimate: 0 bytes, allocs estimate: 0.

julia> @benchmark mul!($C, $A, $B)
BenchmarkTools.Trial: 6 samples with 1 evaluation.
 Range (min … max):  919.987 ms … 941.372 ms  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     930.102 ms               ┊ GC (median):    0.00%
 Time  (mean ± σ):   929.516 ms ±   8.021 ms  ┊ GC (mean ± σ):  0.00% ± 0.00%

  █   █                 █            █   █                    █  
  █▁▁▁█▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁█▁▁▁▁▁▁▁▁▁▁▁▁█▁▁▁█▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁█ ▁
  920 ms           Histogram: frequency by time          941 ms <

 Memory estimate: 25.56 KiB, allocs estimate: 10.
```

`mymulX!` correspond to (X=1) outer loop order `j in BxN, i in AxM`, (X=2) outer loop order `i in AxM, j in BxN`, (X=3) your triple loop but with the B factor hoisted out. Inner loops are all marked with `@simd`.

Reply:
What's the size?
How does the performance compare with `copyto!(C, (B'*A')')`?

I think it's worth taking this performance hit, but if we really want to avoid the regression, we could create just a single tile for the destination array, copying it over periodically.
E.g., the tile could be a vector that we use to fill one row at a time.

These are far from optimal implementations, but the idea is to try and get as much for as little as possible.

Reply:
It's Float32 target, and both A and B are Float64. The `copyto!` is faaaaar better than anything else!

Reply:
With the latest commit, the both-transposed or both-adjoint case is down to almost half of the current `mul!`.

Reply:
BTW, `copyto!(C, (B'*A')')` is not a fair comparison (at least for my test case), because both A and B have same eltype, so `B'*A'` runs by BLAS first.